lib/rpc/singleJsonRpcProvider.ts

```diff
@@ -0,0 +1,460 @@
+    this.recordPerfBeforeCall(startTime)
+    let callSucceed = true
+    return super
+      .resolveName(name)
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('resolveName')
+      })
+  }
+
+  override sendTransaction(signedTransaction: string | Promise<string>): Promise<TransactionResponse> {
+    const startTime = Date.now()
+    this.recordPerfBeforeCall(startTime)
+    let callSucceed = true
+    return super
+      .sendTransaction(signedTransaction)
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('sendTransaction')
+      })
+  }
+
+  override waitForTransaction(
+    transactionHash: string,
+    confirmations?: number,
+    timeout?: number
+  ): Promise<TransactionReceipt> {
+    const startTime = Date.now()
+    this.recordPerfBeforeCall(startTime)
+    let callSucceed = true
+    return super
+      .waitForTransaction(transactionHash, confirmations, timeout)
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('waitForTransaction')
+      })
+  }
+
+  override call(transaction: Deferrable<TransactionRequest>, blockTag?: BlockTag | Promise<BlockTag>): Promise<string> {
+    const startTime = Date.now()
+    this.recordPerfBeforeCall(startTime)
+    let callSucceed = true
+    return super
+      .call(transaction, blockTag)
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('waitForTransaction')
+      })
+  }
+
+  override send(method: string, params: Array<any>): Promise<any> {
+    const startTime = Date.now()
+    this.recordPerfBeforeCall(startTime)
+    let callSucceed = true
+    return super
+      .send(method, params)
+      .then((response) => {
+        return response
+      })
+      .catch((error) => {
+        callSucceed = false
+        throw error
+      })
+      .finally(() => {
+        const endTime = Date.now()
+        this.recordPerfAfterCall(startTime, endTime, callSucceed)
+        this.checkLastCallPerformance('send')
+      })
+  }
```

Comment:
I think we don't need to override all the JSON RPC methods, because many of them are not invoked in prod today.

Originally we decided to override every JSON RPC method in [InstrumentedEVMProvider](https://github.com/Uniswap/routing-api/blob/main/lib/handlers/evm/provider/InstrumentedEVMProvider.ts), so that we can see which methods are being invoked in routing-api runtime, although we already knew some methods are not being invoked like `getLogs`, because we don't have a use case to inspect ETH logs.

You can use the [AWS metrics](https://us-east-2.console.aws.amazon.com/cloudwatch/home?region=us-east-2#metricsV2?graph=~(metrics~(~(~'Uniswap~'RPC_INFURA_1_RESOLVENAME_REQUESTED~'Service~'RoutingAPI)~(~'.~'RPC_INFURA_1_GETNETWORK_REQUESTED~'.~'.)~(~'.~'RPC_INFURA_1_ESTIMATEGAS_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_1_CALL_LATENCY~'.~'.)~(~'.~'RPC_INFURA_1_ESTIMATEGAS_REQUESTED~'.~'.)~(~'.~'RPC_INFURA_1_ESTIMATEGAS_FAILURE~'.~'.)~(~'.~'RPC_INFURA_1_CALL_FAILURE~'.~'.)~(~'.~'RPC_INFURA_1_GETBLOCKNUMBER_REQUESTED~'.~'.)~(~'.~'RPC_INFURA_1_CALL_REQUESTED~'.~'.)~(~'.~'RPC_INFURA_1_CALL_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_1_GETNETWORK_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_1_GETBLOCKNUMBER_FAILURE~'.~'.)~(~'.~'RPC_INFURA_1_GETBLOCKNUMBER_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_1_RESOLVENAME_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_43114_GETNETWORK_REQUESTED~'.~'.)~(~'.~'RPC_INFURA_43114_GETNETWORK_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_43114_GETBLOCKNUMBER_FAILURE~'.~'.)~(~'.~'RPC_INFURA_43114_CALL_LATENCY~'.~'.)~(~'.~'RPC_INFURA_43114_RESOLVENAME_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_43114_CALL_FAILURE~'.~'.)~(~'.~'RPC_INFURA_43114_GETBLOCKNUMBER_REQUESTED~'.~'.)~(~'.~'RPC_INFURA_43114_GETGASPRICE_FAILURE~'.~'.)~(~'.~'RPC_INFURA_43114_GETGASPRICE_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_43114_GETBLOCKNUMBER_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_43114_RESOLVENAME_REQUESTED~'.~'.)~(~'.~'RPC_INFURA_43114_GETGASPRICE_REQUESTED~'.~'.)~(~'.~'RPC_INFURA_43114_CALL_SUCCESS~'.~'.)~(~'.~'RPC_INFURA_43114_CALL_REQUESTED~'.~'.))~view~'timeSeries~stacked~false~region~'us-east-2~start~'-PT1H~end~'P0D~stat~'SampleCount~period~300)&query=~'*7bUniswap*2cService*7d*20RPC_INFURA_43114) to see which methods are invoked today. For example, I want to check mainnet and avalanche, so I put prefix `RPC_INFURA_1` and `RPC_INFURA_43114`, and see:

![Screenshot 2024-01-10 at 11 11 24 AM](https://github.com/Uniswap/routing-api/assets/91580504/09579332-cb28-41d8-9101-d85add30af82)

In fact, the [Routing Dashboard](https://us-east-2.console.aws.amazon.com/cloudwatch/home?region=us-east-2#dashboards/dashboard/RoutingDashboard) only show the JSON RPC methods that are invoked at the bottom of the page, because majority of the JSON RPC methods are not invoked.